### ISSUE TITLE
spsolve error with int64

### ISSUE THREAD

**Issue by juanjimenez** [scipy/scipy#6177]
csr_matrix object creation:
M = csr_matrix((m,m))
M.indices = array([0,1,2, what-so-ever])

B = array([1,2,3...up to m elements])

x = spsolve(M,B) 
everything works in 32 bit platforms  but in 64 bit os systems it does not work. The problem is that M.indices should be 'int32' ctype but by default in the 64 bit system it is 'int64' 
the error reported by spsolve is:
TypeError: colind and rowptr must be of type cint  (somewhat cryptic, in my opinion)

if you define,
M.indices = array([0,1,2, what-so-ever],dtype = 'int32')
everything appears to work again in 64bit systems.

Please consider to document this or to automatically cast the integer types..

Thanks 

Juan Jiménez

**Comment by perimosocordiae** [comment]
I can't reproduce this on my Ubuntu 64-bit machine on Scipy master. What scipy version are you using?

**Comment by rgommers** [comment]
Could be that @juanjimenez is using an old version of scipy. There were a lot of fixes for 64-bit indices in 2014.

@juanjimenez if this is for a recent scipy version, please reopen this issue and provide a standalone example that reproduces the ``TypeError``. Closing now, because as is this bug report isn't actionable.

**Comment by rvcristiand** [comment]
I have the same problem. I made a matrix csr, a array vector and I can't use the spsolve function. Do you know what can I do for this problem ?

**Comment by WarrenWeckesser** [comment]
@rvcristiand What version of scipy are you using?   If you don't know, you can check with:

    import scipy
    print(scipy.__version__)

**Comment by rvcristiand** [comment]
I have the 0.17.0 version.

**Comment by juanjimenez** [comment]
Hello,

Attached to this e-mail, I'm sending you an example of the problem I found
weeks ago and how I went around it.

In the mean time, I have reinstalled the OS of my computer (stable debian ):

Debian 8.5 jessie
linux 3.16.0-4-amd64

I use the Spyder 2.3.1 ,  the version with can with debian stable
distribution, and the Numpy and Scipy version,
Python 2.7.9 64 bits
NumPy 1.8.2
SciPy 0.14.0

... and I can no longer reproduce the problem. This is the reason why I
didn't send an example of the  problem before.
I think Ralf is right. Probably, I was using an old version of scipy.

Regards,

Juan



2016-07-14 19:38 GMT+02:00 rvcristiand <notifications@github.com>:

> I have the 0.17.0 version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/issues/6177#issuecomment-232737089>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ACdBvnrzaSJUnop7VV7Anz7a-G6FWaBSks5qVnQmgaJpZM4IiaF_>
> .
>

# -*- coding: utf-8 -*-
"""
Created on Sun Jun 19 19:22:36 2016
Example to show the bug when using spsolve to solve a system with a sparse
 coefficient matrix...
@author: juan
"""
import numpy as np
from scipy.sparse import csr_matrix
from scipy.sparse.linalg import spsolve

#csr sparse matrix
M = csr_matrix((10,10)) 
    
#pointer to the non-null elements of M 
M.indptr[1:3] = [4,8]
M.indptr[3:-2] = np.arange(14,14+6*(M.indptr.size-5),6)
M.indptr[-2:] = [M.indptr[-3] + 4, M.indptr[-3] + 8]      
    
#Indices of the non-null elements
###############################################################################    
#This asignation didn't work in my 64b computer: 
#M.indices= np.array([0,1,2,3,0,1,2,3,0,1,2,3,4,5,0,1,2,3,4,5])

# spsolver complained: M.indices values should be int32 type
#
#Then, I declared the type explicitly 
M.indices= np.array([0,1,2,3,0,1,2,3,0,1,2,3,4,5,0,1,2,3,4,5],\
dtype = 'int32')

#(please notice -> -> dtype = 'int32')   
#and my program worked again...
###############################################################################
 
#I define the remaining M.indices. (I supose that, once the type has been defined,
# if you add new values of other type the will be cast to int32...)
 
 
for i in range(4,M.indptr.size-4,2):
      M.indices= np.concatenate((M.indices,i-2 + M.indices[8:20]))
M.indices= np.concatenate((M.indices, i + M.indices[0:8]))
# probably, it is not the most-elegant-python's-code-you-have-ever-seen ;)
# The idea is that for rows 1 and 2, no-zero elements are only the first four.
# for rows -2 and -1 non-zero elements are only the last four.
#the remaining rows have six non-zero elements; each two rows, the six elements
#are shifted two columns to the right.
# well, perhaps is better to see the result... running the module

# then , I fill the non-zero elements. Values are irrelevant, it's just to
#show results...
M.data = np.arange(1,2*M.indices.size,2)

print M.toarray()

a = np.ones(10)
print spsolve(M,a)  

#end of module

**Comment by rgommers** [comment]
Thanks for confirming Juan

**Comment by TheIdealis** [comment]
I recently switched to conda, and now I've got the problem too.
- scipy.__version__:  '0.19.1'
- numpy.__version__: '1.13.1'

As Juan has shown before, if I don't explicitly define the datatype

```python
from scipy.sparse import csr_matrix   
from scipy.sparse.linalg import spsolve   
import numpy as np   
M = csr_matrix((10,10))   
M.indices = np.array([0,1,2])   
B = np.ones(10)   
x = spsolve(M,B)
```

I get the error:

```python
~/.conda/envs/basic_python/lib/python3.6/site-packages/scipy/sparse/linalg/dsolve/linsolve.py in spsolve(A, b, permc_spec, use_umfpack)
    157             options = dict(ColPerm=permc_spec)
    158             x, info = _superlu.gssv(N, A.nnz, A.data, A.indices, A.indptr,
--> 159                                     b, flag, options=options)
    160             if info != 0:
    161                 warn("Matrix is exactly singular", MatrixRankWarning)

TypeError: colind and rowptr must be of type cint
```
or for UMFPACK:

```python
~/.conda/envs/basic_python/lib/python3.6/site-packages/scipy/sparse/linalg/dsolve/linsolve.py in spsolve(A, b, permc_spec, use_umfpack)
    143         umf = umfpack.UmfpackContext(_get_umf_family(A))
    144         x = umf.linsolve(umfpack.UMFPACK_A, A, b_vec,
--> 145                          autoTranspose=True)
    146     else:
    147         if b_is_vector and b_is_sparse:

~/.conda/envs/basic_python/lib/python3.6/site-packages/scikits/umfpack/umfpack.py in linsolve(self, sys, mtx, rhs, autoTranspose)
    751 
    752         if self._numeric is None:
--> 753             self.numeric(mtx)
    754         else:
    755             if self.mtx is not mtx:

~/.conda/envs/basic_python/lib/python3.6/site-packages/scikits/umfpack/umfpack.py in numeric(self, mtx)
    537 
    538         if self._symbolic is None:
--> 539             self.symbolic(mtx)
    540 
    541         indx = self._getIndx(mtx)

~/.conda/envs/basic_python/lib/python3.6/site-packages/scikits/umfpack/umfpack.py in symbolic(self, mtx)
    497         self.free_symbolic()
    498 
--> 499         indx = self._getIndx(mtx)
    500 
    501         if not assumeSortedIndices:

~/.conda/envs/basic_python/lib/python3.6/site-packages/scikits/umfpack/umfpack.py in _getIndx(self, mtx)
    476             if (indx.dtype != np.dtype('l')) \
    477                    or mtx.indptr.dtype != np.dtype('l'):
--> 478                 raise ValueError('matrix must have long indices')
    479 
    480         if self.isReal:

ValueError: matrix must have long indices
```

**Comment by sonsongithub** [comment]
Why this issue would be closed?
Now, I've got the same problem, too.

**Comment by perimosocordiae** [comment]
You may be hitting https://github.com/scipy/scipy/issues/18603, which was recently fixed and will be part of the 1.12 release.
